libyul/optimiser/UnusedFunctionsCommon.h

```diff
@@ -0,0 +1,171 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+
+#include <liblangutil/SourceLocation.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <variant>
+
+namespace solidity::yul
+{
+
+template<typename T>
+std::vector<T> applyBooleanMask(std::vector<T> const& _vec, std::vector<bool> const& _mask)
+{
+	yulAssert(_vec.size() == _mask.size(), "");
+
+	std::vector<T> ret;
+
+	for (size_t i = 0; i < _mask.size(); ++i)
+		if (_mask[i])
+			ret.push_back(_vec[i]);
+
+	return ret;
+}
+
+/// Find functions whose arguments are not used in its body. Also, find functions whose body
+/// satisfies a heuristic about pruning.
+bool wasPruned(FunctionDefinition const& _f, Dialect const& _dialect)
+{
+	// We skip the function body if it
+	// 1. is empty, or
+	// 2. is a single statement that is an assignment statement whose value is a non-builtin
+	//    function call, or
+	// 3. is a single expression-statement that is a non-builtin function call.
+	// The above cases are simple enough so that the inliner alone can remove the parameters.
+	if (_f.body.statements.empty())
+		return true;
+	if (_f.body.statements.size() == 1)
+	{
+		Statement const& e = _f.body.statements[0];
+		if (std::holds_alternative<Assignment>(e))
+		{
+			if (std::holds_alternative<FunctionCall>(*std::get<Assignment>(e).value))
+			{
+				FunctionCall c = std::get<FunctionCall>(*std::get<Assignment>(e).value);
+				if (!_dialect.builtin(c.functionName.name))
+					return true;
+			}
+		}
+		else if (std::holds_alternative<ExpressionStatement>(e))
+			if (std::holds_alternative<FunctionCall>(std::get<ExpressionStatement>(e).expression))
+			{
+				FunctionCall c = std::get<FunctionCall>(std::get<ExpressionStatement>(e).expression);
+				if (!_dialect.builtin(c.functionName.name))
+					return true;
+			}
+	}
+	return false;
+}
+
+FunctionDefinition createReplacement(
+	FunctionDefinition& _old,
+	std::map<YulString, std::vector<bool>> const& _unusedParameters,
+	std::map<YulString, std::vector<bool>> const& _unusedReturnVariables,
+	NameDispenser& _nameDispenser,
+	std::map<YulString, YulString> _inverseTranslations
```

Comment:
Same here, we only need old and new name, right?

Reply:
This can be avoided. This should be a `const` reference. I'll fix that.